src/graphql/rest-authorization.guard.ts

```diff
@@ -0,0 +1,79 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { Reflector } from '@nestjs/core';
+import { Request, Response } from 'express'; // Use Request from 'express'
+import { expressjwt } from 'express-jwt';
+import { expressJwtSecret, GetVerificationKey } from 'jwks-rsa';
+import { PERMISSION_SCOPES, Role, ROLES_KEY } from 'src/util/constants';
+import { promisify } from 'util';
+
+interface AuthenticatedRequest extends Request {
+  auth?: {
+    permissions?: string[];
+    [key: string]: any;
+  };
+}
+
+@Injectable()
+export class RestAuthorizationGuard implements CanActivate {
+  private AUTH0_AUDIENCE: string;
+  private AUTH0_DOMAIN: string;
+
+  constructor(
+    private configService: ConfigService,
+    private reflector: Reflector,
+  ) {
+    this.AUTH0_AUDIENCE = this.configService.get('AUTH0_AUDIENCE');
+    this.AUTH0_DOMAIN = this.configService.get('AUTH0_DOMAIN');
+  }
+
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const req = context.switchToHttp().getRequest<AuthenticatedRequest>();
+    const res = context.switchToHttp().getResponse<Response>();
+
+    const checkJWT = promisify(
+      expressjwt({
+        secret: expressJwtSecret({
+          cache: true,
+          rateLimit: true,
+          jwksRequestsPerMinute: 5,
+          jwksUri: `${this.AUTH0_DOMAIN}.well-known/jwks.json`,
+        }) as GetVerificationKey,
+        audience: this.AUTH0_AUDIENCE,
+        issuer: this.AUTH0_DOMAIN,
+        algorithms: ['RS256'],
+      }),
+    );
+
+    try {
+      await checkJWT(req, res);
+
+      const requiredRoles = this.reflector.getAllAndOverride<Role[]>(
+        ROLES_KEY,
+        [context.getHandler(), context.getClass()],
+      );
+
+      const scopeRules = req?.auth?.permissions as string[];
+      if (requiredRoles) {
+        if (!scopeRules?.length) return false;
+
+        const [requiredRole] = requiredRoles;
+
+        const hasAccess = scopeRules.every((scopeType) =>
+          PERMISSION_SCOPES[requiredRole].includes(scopeType),
+        );
+
+        return hasAccess;
+      }
+
+      return true;
+    } catch (err) {
+      throw new UnauthorizedException(err);
+    }
+  }
+}
```

Comment:
Great work! could you move to outside a graqhql folder ? and remove rest- prefix 

For now please follow this estructure:

├── shared/
│   ├── decorators/
│   │   ├── current-user.decorator.ts
│   │   └── roles.decorator.ts
│   └── guards/
│       ├── auth.guard.ts
│       ├── roles.guard.ts
│       └── authorization.guard.ts

Reply:
We could type this?

` [key: string]: any;`

Reply:
Interessant avoid use AS on typescript code:

```
const SECRET_OPTIONS: GetVerificationKey = { 
          cache: true,
          rateLimit: true,
          jwksRequestsPerMinute: 5,
          jwksUri: `${this.AUTH0_DOMAIN}.well-known/jwks.json`,
}

const checkJWT = promisify(
      expressjwt({
        secret: expressJwtSecret(SECRET_OPTIONS),
        audience: this.AUTH0_AUDIENCE,
        issuer: this.AUTH0_DOMAIN,
        algorithms: ['RS256'],
      }),
    );

```

Reply:
`const scopeRules = req?.auth?.permissions as string[];`

We need type auth.permessions correctly and avoid create a fragile type using as to override.

Reply:
Sure, i will look how to do this properly

Reply:
https://github.com/auth0/node-jwks-rsa/pull/412
update of jwks-rsa and express-jwt still doesn't worked the expressJwtSecret function;

As they recomendate: https://github.com/auth0/express-jwt/issues/288, i will keep the "as GetVerificationKey"

Also, removed the [key: string]: any; it wasn't necessary